// @kubiklabs/wasmkit-ast-types/main/wasm.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.createWasmQueryMethod = exports.createWasmExecMethod = exports.createQueryInterface = exports.createQueryClass = exports.createPropertyFunctionWithObjectParamsForExec = exports.createPropertyFunctionWithObjectParams = exports.createExecuteInterface = exports.createExecuteClass = void 0;
var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));
var _toConsumableArray2 = _interopRequireDefault(require("@babel/runtime/helpers/toConsumableArray"));
var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));
var t = _interopRequireWildcard(require("@babel/types"));
var _case = require("case");
var _utils = require("./utils");
var _babel = require("./utils/babel");
var _types2 = require("./utils/types");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { "default": e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && Object.prototype.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n["default"] = e, t && t.set(e, n), n; }
var createWasmQueryMethod = exports.createWasmQueryMethod = /*#__PURE__*/function () {
  var _ref = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee(jsonschema, definitions) {
    var _jsonschema$propertie;
    var underscoreName, methodName, responseType, properties, obj, args, actionArg;
    return _regenerator["default"].wrap(function _callee$(_context) {
      while (1) switch (_context.prev = _context.next) {
        case 0:
          underscoreName = Object.keys(jsonschema.properties)[0];
          methodName = (0, _case.camel)(underscoreName);
          responseType = "any";
          properties = (_jsonschema$propertie = jsonschema.properties[underscoreName].properties) !== null && _jsonschema$propertie !== void 0 ? _jsonschema$propertie : {};
          _context.next = 6;
          return (0, _types2.createTypedObjectParams)(jsonschema.properties[underscoreName], definitions);
        case 6:
          obj = _context.sent;
          args = Object.keys(properties).map(function (prop) {
            return t.objectProperty(t.identifier(prop), t.identifier((0, _case.camel)(prop)), false, true);
          });
          actionArg = t.objectProperty(t.identifier(underscoreName), t.objectExpression(args));
          return _context.abrupt("return", t.classProperty(t.identifier(methodName), (0, _utils.arrowFunctionExpression)(obj ? [obj] : [], t.blockStatement([t.returnStatement(t.callExpression(t.memberExpression(t.thisExpression(), t.identifier('queryMsg')), [t.objectExpression([actionArg])]))]), t.tsTypeAnnotation(t.tsTypeReference(t.identifier('Promise'), t.tsTypeParameterInstantiation([t.tSTypeReference(t.identifier(responseType))]))), true)));
        case 10:
        case "end":
          return _context.stop();
      }
    }, _callee);
  }));
  return function createWasmQueryMethod(_x, _x2) {
    return _ref.apply(this, arguments);
  };
}();
var createWasmExecMethod = exports.createWasmExecMethod = /*#__PURE__*/function () {
  var _ref2 = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee2(jsonschema, definitions) {
    var _jsonschema$propertie2;
    var underscoreName, methodName, properties, obj, changeConstParamNames, _i, _Object$keys, prop, args, accountVar, customFeesVar, memoVar, transferAmountVar, constantParams;
    return _regenerator["default"].wrap(function _callee2$(_context2) {
      while (1) switch (_context2.prev = _context2.next) {
        case 0:
          underscoreName = Object.keys(jsonschema.properties)[0];
          methodName = (0, _case.camel)(underscoreName);
          properties = (_jsonschema$propertie2 = jsonschema.properties[underscoreName].properties) !== null && _jsonschema$propertie2 !== void 0 ? _jsonschema$propertie2 : {};
          _context2.next = 5;
          return (0, _types2.createTypedObjectParams)(jsonschema.properties[underscoreName], definitions);
        case 5:
          obj = _context2.sent;
          changeConstParamNames = false;
          for (_i = 0, _Object$keys = Object.keys(properties); _i < _Object$keys.length; _i++) {
            prop = _Object$keys[_i];
            if (prop === 'memo' || prop === 'account' || prop === 'customFees' || prop === 'transferAmount') {
              changeConstParamNames = true;
            }
          }
          args = Object.keys(properties).map(function (prop) {
            return t.objectProperty(t.identifier(prop), t.identifier((0, _case.camel)(prop)), false, prop === (0, _case.camel)(prop));
          });
          accountVar = changeConstParamNames ? 'txnAccount' : 'account';
          customFeesVar = changeConstParamNames ? 'txnCustomFees' : 'customFees';
          memoVar = changeConstParamNames ? 'txnMemo' : 'memo';
          transferAmountVar = changeConstParamNames ? 'txnTransferAmount' : 'transferAmount';
          constantParams = t.objectPattern([t.objectProperty(t.identifier(accountVar), t.identifier(accountVar), false, true), t.objectProperty(t.identifier(customFeesVar), t.identifier(customFeesVar), false, true), t.objectProperty(t.identifier(memoVar), t.identifier(memoVar), false, true), t.objectProperty(t.identifier(transferAmountVar), t.identifier(transferAmountVar), false, true)]);
          constantParams.typeAnnotation = t.tsTypeAnnotation(t.tsTypeLiteral([t.tSPropertySignature(t.identifier(accountVar), t.tsTypeAnnotation(t.tsTypeReference(t.identifier('wasmKitTypes.UserAccount')))), t.tSPropertySignature(t.identifier("".concat(customFeesVar, "?")), t.tsTypeAnnotation(t.tsTypeReference(t.identifier('wasmKitTypes.TxnStdFee')))), t.tSPropertySignature(t.identifier("".concat(memoVar, "?")), t.tsTypeAnnotation(t.tsStringKeyword())), t.tSPropertySignature(t.identifier("".concat(transferAmountVar, "?")), t.tsTypeAnnotation((0, _babel.tsTypeOperator)(t.tsArrayType(t.tsTypeReference(t.identifier('Coin'))), 'readonly')))]));
          return _context2.abrupt("return", t.classProperty(t.identifier(methodName), (0, _utils.arrowFunctionExpression)(obj ? [
          // props
          constantParams, obj] : [constantParams], t.blockStatement([t.returnStatement(t.awaitExpression(t.callExpression(t.memberExpression(t.thisExpression(), t.identifier('executeMsg')), [t.objectExpression([t.objectProperty(t.identifier(underscoreName), t.objectExpression((0, _toConsumableArray2["default"])(args)))]), t.identifier(accountVar), t.identifier(customFeesVar), t.identifier(memoVar), t.identifier(transferAmountVar)])))]),
          // return type
          t.tsTypeAnnotation(t.tsTypeReference(t.identifier('Promise'), t.tsTypeParameterInstantiation([t.tSTypeReference(t.identifier('any'))]))), true)));
        case 16:
        case "end":
          return _context2.stop();
      }
    }, _callee2);
  }));
  return function createWasmExecMethod(_x3, _x4) {
    return _ref2.apply(this, arguments);
  };
}();
var createQueryClass = exports.createQueryClass = /*#__PURE__*/function () {
  var _ref3 = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee4(className, implementsClassName, extendsClassName, queryMsg, skipSchemaErrors) {
    var propertyNames, bindings, methods;
    return _regenerator["default"].wrap(function _callee4$(_context4) {
      while (1) switch (_context4.prev = _context4.next) {
        case 0:
          propertyNames = (0, _utils.getMessageProperties)(queryMsg).map(function (method) {
            var _Object$keys2;
            return (_Object$keys2 = Object.keys(method.properties)) === null || _Object$keys2 === void 0 ? void 0 : _Object$keys2[0];
          }).filter(Boolean);
          bindings = propertyNames.map(_case.camel).map(_utils.bindMethod);
          _context4.next = 4;
          return Promise.all((0, _utils.getMessageProperties)(queryMsg).map( /*#__PURE__*/function () {
            var _ref4 = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee3(schema) {
              return _regenerator["default"].wrap(function _callee3$(_context3) {
                while (1) switch (_context3.prev = _context3.next) {
                  case 0:
                    _context3.prev = 0;
                    _context3.next = 3;
                    return createWasmQueryMethod(schema, queryMsg.definitions);
                  case 3:
                    return _context3.abrupt("return", _context3.sent);
                  case 6:
                    _context3.prev = 6;
                    _context3.t0 = _context3["catch"](0);
                    if (!skipSchemaErrors) {
                      _context3.next = 12;
                      break;
                    }
                    return _context3.abrupt("return", null);
                  case 12:
                    throw _context3.t0;
                  case 13:
                  case "end":
                    return _context3.stop();
                }
              }, _callee3, null, [[0, 6]]);
            }));
            return function (_x10) {
              return _ref4.apply(this, arguments);
            };
          }()).filter(function (method) {
            return method !== null;
          }));
        case 4:
          methods = _context4.sent;
          return _context4.abrupt("return", t.exportNamedDeclaration((0, _utils.classDeclaration)(className, [
          // constructor
          t.classMethod('constructor', t.identifier('constructor'), [(0, _utils.typedIdentifier)('contractName', t.tsTypeAnnotation(t.tsStringKeyword())), (0, _utils.typedIdentifier)('instantiateTag?', t.tsTypeAnnotation(t.tsStringKeyword()))], t.blockStatement([t.expressionStatement(t.callExpression(t["super"](), [t.identifier('contractName'), t.identifier('instantiateTag')]))].concat((0, _toConsumableArray2["default"])(bindings))))].concat((0, _toConsumableArray2["default"])(methods)), [t.tSExpressionWithTypeArguments(t.identifier(implementsClassName))], extendsClassName ? t.identifier(extendsClassName) : null)));
        case 6:
        case "end":
          return _context4.stop();
      }
    }, _callee4);
  }));
  return function createQueryClass(_x5, _x6, _x7, _x8, _x9) {
    return _ref3.apply(this, arguments);
  };
}();
var createExecuteClass = exports.createExecuteClass = /*#__PURE__*/function () {
  var _ref5 = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee6(className, implementsClassName, extendsClassName, execMsg, contractName, skipSchemaErrors) {
    var propertyNames, bindings, methods, blockStmt;
    return _regenerator["default"].wrap(function _callee6$(_context6) {
      while (1) switch (_context6.prev = _context6.next) {
        case 0:
          propertyNames = (0, _utils.getMessageProperties)(execMsg).map(function (method) {
            var _Object$keys3;
            return (_Object$keys3 = Object.keys(method.properties)) === null || _Object$keys3 === void 0 ? void 0 : _Object$keys3[0];
          }).filter(Boolean);
          bindings = propertyNames.map(_case.camel).map(_utils.bindMethod);
          _context6.next = 4;
          return Promise.all((0, _utils.getMessageProperties)(execMsg).map( /*#__PURE__*/function () {
            var _ref6 = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee5(schema) {
              return _regenerator["default"].wrap(function _callee5$(_context5) {
                while (1) switch (_context5.prev = _context5.next) {
                  case 0:
                    _context5.prev = 0;
                    _context5.next = 3;
                    return createWasmExecMethod(schema, execMsg.definitions);
                  case 3:
                    return _context5.abrupt("return", _context5.sent);
                  case 6:
                    _context5.prev = 6;
                    _context5.t0 = _context5["catch"](0);
                    if (!skipSchemaErrors) {
                      _context5.next = 12;
                      break;
                    }
                    return _context5.abrupt("return", null);
                  case 12:
                    throw _context5.t0;
                  case 13:
                  case "end":
                    return _context5.stop();
                }
              }, _callee5, null, [[0, 6]]);
            }));
            return function (_x17) {
              return _ref6.apply(this, arguments);
            };
          }()).filter(function (method) {
            return method !== null;
          }));
        case 4:
          methods = _context6.sent;
          blockStmt = [];
          if (extendsClassName) {
            blockStmt.push(t.expressionStatement(t.callExpression(t["super"](), [t.stringLiteral(contractName), t.identifier('instantiateTag')])));
          }
          [].push.apply(blockStmt, (0, _toConsumableArray2["default"])(bindings));
          return _context6.abrupt("return", t.exportNamedDeclaration((0, _utils.classDeclaration)(className, [
          // constructor
          t.classMethod('constructor', t.identifier('constructor'), [(0, _utils.typedIdentifier)('instantiateTag?', t.tsTypeAnnotation(t.tsStringKeyword()))], t.blockStatement(blockStmt))].concat((0, _toConsumableArray2["default"])(methods)), [t.tSExpressionWithTypeArguments(t.identifier(implementsClassName))], extendsClassName ? t.identifier(extendsClassName) : null)));
        case 9:
        case "end":
          return _context6.stop();
      }
    }, _callee6);
  }));
  return function createExecuteClass(_x11, _x12, _x13, _x14, _x15, _x16) {
    return _ref5.apply(this, arguments);
  };
}();
var createQueryInterface = exports.createQueryInterface = /*#__PURE__*/function () {
  var _ref7 = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee8(className, queryMsg, skipSchemaErrors) {
    var methods;
    return _regenerator["default"].wrap(function _callee8$(_context8) {
      while (1) switch (_context8.prev = _context8.next) {
        case 0:
          _context8.next = 2;
          return Promise.all((0, _utils.getMessageProperties)(queryMsg).map( /*#__PURE__*/function () {
            var _ref8 = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee7(jsonschema) {
              var underscoreName, methodName, responseType;
              return _regenerator["default"].wrap(function _callee7$(_context7) {
                while (1) switch (_context7.prev = _context7.next) {
                  case 0:
                    underscoreName = Object.keys(jsonschema.properties)[0];
                    methodName = (0, _case.camel)(underscoreName);
                    responseType = "any";
                    _context7.prev = 3;
                    _context7.next = 6;
                    return createPropertyFunctionWithObjectParams(methodName, responseType, jsonschema.properties[underscoreName], queryMsg.definitions);
                  case 6:
                    return _context7.abrupt("return", _context7.sent);
                  case 9:
                    _context7.prev = 9;
                    _context7.t0 = _context7["catch"](3);
                    if (!skipSchemaErrors) {
                      _context7.next = 15;
                      break;
                    }
                    return _context7.abrupt("return", null);
                  case 15:
                    throw _context7.t0;
                  case 16:
                  case "end":
                    return _context7.stop();
                }
              }, _callee7, null, [[3, 9]]);
            }));
            return function (_x21) {
              return _ref8.apply(this, arguments);
            };
          }()).filter(function (method) {
            return method !== null;
          }));
        case 2:
          methods = _context8.sent;
          return _context8.abrupt("return", t.exportNamedDeclaration(t.tsInterfaceDeclaration(t.identifier(className), null, [], t.tSInterfaceBody((0, _toConsumableArray2["default"])(methods)))));
        case 4:
        case "end":
          return _context8.stop();
      }
    }, _callee8);
  }));
  return function createQueryInterface(_x18, _x19, _x20) {
    return _ref7.apply(this, arguments);
  };
}();
var createExecuteInterface = exports.createExecuteInterface = /*#__PURE__*/function () {
  var _ref9 = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee10(className, extendsClassName, execMsg, skipSchemaErrors) {
    var methods, extendsAst;
    return _regenerator["default"].wrap(function _callee10$(_context10) {
      while (1) switch (_context10.prev = _context10.next) {
        case 0:
          _context10.next = 2;
          return Promise.all((0, _utils.getMessageProperties)(execMsg).map( /*#__PURE__*/function () {
            var _ref10 = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee9(jsonschema) {
              var underscoreName, methodName;
              return _regenerator["default"].wrap(function _callee9$(_context9) {
                while (1) switch (_context9.prev = _context9.next) {
                  case 0:
                    underscoreName = Object.keys(jsonschema.properties)[0];
                    methodName = (0, _case.camel)(underscoreName);
                    _context9.prev = 2;
                    _context9.next = 5;
                    return createPropertyFunctionWithObjectParamsForExec(methodName, 'any', jsonschema.properties[underscoreName], execMsg.definitions);
                  case 5:
                    return _context9.abrupt("return", _context9.sent);
                  case 8:
                    _context9.prev = 8;
                    _context9.t0 = _context9["catch"](2);
                    if (!skipSchemaErrors) {
                      _context9.next = 14;
                      break;
                    }
                    return _context9.abrupt("return", null);
                  case 14:
                    throw _context9.t0;
                  case 15:
                  case "end":
                    return _context9.stop();
                }
              }, _callee9, null, [[2, 8]]);
            }));
            return function (_x26) {
              return _ref10.apply(this, arguments);
            };
          }()).filter(function (method) {
            return method !== null;
          }));
        case 2:
          methods = _context10.sent;
          extendsAst = extendsClassName ? [t.tSExpressionWithTypeArguments(t.identifier(extendsClassName))] : [];
          return _context10.abrupt("return", t.exportNamedDeclaration(t.tsInterfaceDeclaration(t.identifier(className), null, extendsAst, t.tSInterfaceBody((0, _toConsumableArray2["default"])(methods)))));
        case 5:
        case "end":
          return _context10.stop();
      }
    }, _callee10);
  }));
  return function createExecuteInterface(_x22, _x23, _x24, _x25) {
    return _ref9.apply(this, arguments);
  };
}();
var createPropertyFunctionWithObjectParams = exports.createPropertyFunctionWithObjectParams = /*#__PURE__*/function () {
  var _ref11 = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee11(methodName, responseType, jsonschema, definitions) {
    var obj, func;
    return _regenerator["default"].wrap(function _callee11$(_context11) {
      while (1) switch (_context11.prev = _context11.next) {
        case 0:
          _context11.next = 2;
          return (0, _types2.createTypedObjectParams)(jsonschema, definitions);
        case 2:
          obj = _context11.sent;
          func = {
            type: 'TSFunctionType',
            typeAnnotation: (0, _utils.promiseTypeAnnotation)(responseType),
            parameters: obj ? [obj] : []
          };
          return _context11.abrupt("return", t.tSPropertySignature(t.identifier(methodName), t.tsTypeAnnotation(func)));
        case 5:
        case "end":
          return _context11.stop();
      }
    }, _callee11);
  }));
  return function createPropertyFunctionWithObjectParams(_x27, _x28, _x29, _x30) {
    return _ref11.apply(this, arguments);
  };
}();
var createPropertyFunctionWithObjectParamsForExec = exports.createPropertyFunctionWithObjectParamsForExec = /*#__PURE__*/function () {
  var _ref12 = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee12(methodName, responseType, jsonschema, definitions) {
    var _jsonschema$propertie3;
    var obj, properties, changeConstParamNames, _i2, _Object$keys4, prop, accountVar, customFeesVar, memoVar, transferAmountVar, fixedParams, func;
    return _regenerator["default"].wrap(function _callee12$(_context12) {
      while (1) switch (_context12.prev = _context12.next) {
        case 0:
          _context12.next = 2;
          return (0, _types2.createTypedObjectParams)(jsonschema, definitions);
        case 2:
          obj = _context12.sent;
          properties = (_jsonschema$propertie3 = jsonschema.properties) !== null && _jsonschema$propertie3 !== void 0 ? _jsonschema$propertie3 : {};
          changeConstParamNames = false;
          for (_i2 = 0, _Object$keys4 = Object.keys(properties); _i2 < _Object$keys4.length; _i2++) {
            prop = _Object$keys4[_i2];
            if (prop === 'memo' || prop === 'account' || prop === 'customFees' || prop === 'transferAmount') {
              changeConstParamNames = true;
            }
          }
          accountVar = changeConstParamNames ? 'txnAccount' : 'account';
          customFeesVar = changeConstParamNames ? 'txnCustomFees' : 'customFees';
          memoVar = changeConstParamNames ? 'txnMemo' : 'memo';
          transferAmountVar = changeConstParamNames ? 'txnTransferAmount' : 'transferAmount';
          fixedParams = t.objectPattern([t.objectProperty(t.identifier(accountVar), t.identifier(accountVar), false, true), t.objectProperty(t.identifier(customFeesVar), t.identifier(customFeesVar), false, true), t.objectProperty(t.identifier(memoVar), t.identifier(memoVar), false, true), t.objectProperty(t.identifier(transferAmountVar), t.identifier(transferAmountVar), false, true)]);
          fixedParams.typeAnnotation = t.tsTypeAnnotation(t.tsTypeLiteral([t.tSPropertySignature(t.identifier(accountVar), t.tsTypeAnnotation(t.tsTypeReference(t.identifier('wasmKitTypes.UserAccount')))), t.tSPropertySignature(t.identifier("".concat(customFeesVar, "?")), t.tsTypeAnnotation(t.tsTypeReference(t.identifier('wasmKitTypes.TxnStdFee')))), t.tSPropertySignature(t.identifier("".concat(memoVar, "?")), t.tsTypeAnnotation(t.tsStringKeyword())), t.tSPropertySignature(t.identifier("".concat(transferAmountVar, "?")), t.tsTypeAnnotation((0, _babel.tsTypeOperator)(t.tsArrayType(t.tsTypeReference(t.identifier('Coin'))), 'readonly')))]));
          func = {
            type: 'TSFunctionType',
            typeAnnotation: (0, _utils.promiseTypeAnnotation)(responseType),
            parameters: obj ? [fixedParams, obj] : [fixedParams]
          };
          return _context12.abrupt("return", t.tSPropertySignature(t.identifier(methodName), t.tsTypeAnnotation(func)));
        case 14:
        case "end":
          return _context12.stop();
      }
    }, _callee12);
  }));
  return function createPropertyFunctionWithObjectParamsForExec(_x31, _x32, _x33, _x34) {
    return _ref12.apply(this, arguments);
  };
}();
