// @kubiklabs/wasmkit/dist/types.js
"use strict";
/* eslint-disable no-use-before-define */
// This file defines the different config types.
//
// For each possible kind of config value, we have two type:
//
// One that ends with UserConfig, which represent the config as
// written in the user's config file.
//
// The other one, with the same name except without the User part, represents
// the resolved value as used during the WasmKit execution.
//
// Note that while many declarations are repeated here (i.e. network types'
// fields), we don't use `extends` as that can interfere with plugin authors
// trying to augment the config types.
// Networks config\
Object.defineProperty(exports, "__esModule", { value: true });
exports.ChainType = void 0;
var ChainType;
(function (ChainType) {
    ChainType[ChainType["Secret"] = 0] = "Secret";
    ChainType[ChainType["Juno"] = 1] = "Juno";
    ChainType[ChainType["Osmosis"] = 2] = "Osmosis";
    ChainType[ChainType["Archway"] = 3] = "Archway";
    ChainType[ChainType["Neutron"] = 4] = "Neutron";
    ChainType[ChainType["Atom"] = 5] = "Atom";
    ChainType[ChainType["Terra"] = 6] = "Terra";
    ChainType[ChainType["Injective"] = 7] = "Injective";
    ChainType[ChainType["Umee"] = 8] = "Umee";
    ChainType[ChainType["Nibiru"] = 9] = "Nibiru";
})(ChainType = exports.ChainType || (exports.ChainType = {}));
